File: Algorithms/6. ZigZag Conversion/python/solution.sort.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s

        rows = [""]*numRows
        cur_row = 0
        going_down = -1

        for c in s:
            rows[cur_row] += c
            if cur_row == 0 or cur_row == numRows-1:
                going_down *= -1
            cur_row += going_down

        return "".join(rows)

File: Algorithms/6. ZigZag Conversion/javascript/solution.sort.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
    if (numRows === 1) {
        return s;
    }

    const rows = new Array(numRows).fill('');
    let curRow = 0,
        goingDown = -1;

    for (const c of s) {
        rows[curRow] += c;
        if (curRow === 0 || curRow === numRows - 1) {
            goingDown *= -1;
        }
        curRow += goingDown;
    }

    return rows.join('');
};
